### PR TITLE
[AppService] Fix #12251 app settings race condition during zip deploy

### LINKED ISSUE #12251: Fail to deploy using az functionapp deployment source config-zip
**CLI version**
2.0.81

**Describe the bug**
When deploying a local function app content to Azure Functions on Linux, the deployment can possibly fails if it requires **changes in app settings**.

There's a race condition here:
1. If the app settings are changed before the deployment is initiated, the build **succeeds**.
2. If the app settings are changed at the same time as the deployment is initiated, the /deployments/latest endpoint will return **5xx** errors due to KuduLite restart.
3. If the app settings are changed after the deployment is started, **the build will be interrupted and marked as failed deployment**.

**To Reproduce**
1. Create a Python Function App in Linux Premium Plan
2. Prepare a simple Python app content, and compress it into a zip file.
3. Run the command `az functionapp deployment source config-zip -g <resource_group> -n <app_name> --src <local_zip_file>`
4. Wait a while and will see the deployment failure error

**Expected behavior**
The deployment should succeed.

### PR DESCRIPTION
**Pull request by Hazhzeng**
Resolves: https://github.com/Azure/azure-cli/issues/12251

**History Notes:**  
[App Service - functionapp] (az functionapp deployment source config-zip) Fix an issue caused by app settings change racing condition during zip deploy, giving 5xx errors during deployment.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [X] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [X] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by Hazhzeng** [comment]
Please review when you have time @ankitkumarr @gzuber

**Comment by yungezz** [comment]
hi @qwordy could you pls help to review? thanks.

**Comment by yonzhan** [comment]
add to S166.

**Comment by qwordy** [comment]
> 
> 
> hi @qwordy could you pls help to review? thanks.

Yeah. it's a large PR.

> 
> 
> hi @qwordy could you pls help to review? thanks.

yeah. it's a large PR.

**Comment by Hazhzeng** [comment]
Hi @qwordy,

Yea sorry for the confusion when you reviewing this PR, let me guide you through with the following steps.

In Azure Functions App, there are two sites: the main site for handling user requests, and the **SCM site for performing the deployment build**. When changing an app setting in Azure CLI, **it takes a while to propagate to the SCM site**. If a build is initiated during app settings propagation, the build will be interrupted.

This main logic change in this PR are as follows:
1. [L422](https://github.com/Azure/azure-cli/pull/12262/files#diff-75a1767a60b87df72982d626136921dbR422) [L457](https://github.com/Azure/azure-cli/pull/12262/files#diff-75a1767a60b87df72982d626136921dbR457) Add a polling blocker and wait for app settings to be completely propagated to the SCM site before initiating a build action.
2. [L800](https://github.com/Azure/azure-cli/pull/12262/files#diff-75a1767a60b87df72982d626136921dbR800) Create a utility for comparing if the app settings in SCM site match the ARM one, for ensuring the propagation is completed.
3. [L820](https://github.com/Azure/azure-cli/pull/12262/files#diff-75a1767a60b87df72982d626136921dbR820) Create a REST request with retry pattern for querying app settings from the SCM site api/settings endpoint.

If you have any further questions, feel free to let me know. Thanks in advance.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 1]
```diff
@@ -380,54 +381,92 @@ def enable_zip_deploy(cmd, resource_group_name, name, src, timeout=None, slot=No
...
             scm_do_build_during_deployment = value in ('true', '1')
         if keyval['name'] == 'WEBSITE_RUN_FROM_PACKAGE':
             website_run_from_package = value
+        if keyval['name'] == 'ENABLE_ORYX_BUILD':
+            enable_oryx_build = value
 
-    if not ((enable_oryx_build is True) and (scm_do_build_during_deployment is True)):
-        logger.warning("Setting ENABLE_ORYX_BUILD to true")
+    if scm_do_build_during_deployment is not True:
         logger.warning("Setting SCM_DO_BUILD_DURING_DEPLOYMENT to true")
```
[the remark this answers is not in the document]
Comment (ankitkumarr) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py]: I think it's logging if the variable is not true, indicating that the Application Setting will now be set to true.

**Comment by qwordy** [comment]
> 
> 
> Hi @qwordy,
> 
> Yea sorry for the confusion when you reviewing this PR, let me guide you through with the following steps.
> 
> In Azure Functions App, there are two sites: the main site for handling user requests, and the **SCM site for performing the deployment build**. When changing an app setting in Azure CLI, **it takes a while to propagate to the SCM site**. If a build is initiated during app settings propagation, the build will be interrupted.
> 
> This main logic change in this PR are as follows:
> 
>     1. [L422](https://github.com/Azure/azure-cli/pull/12262/files#diff-75a1767a60b87df72982d626136921dbR422) [L457](https://github.com/Azure/azure-cli/pull/12262/files#diff-75a1767a60b87df72982d626136921dbR457) Add a polling blocker and wait for app settings to be completely propagated to the SCM site before initiating a build action.
> 
>     2. [L800](https://github.com/Azure/azure-cli/pull/12262/files#diff-75a1767a60b87df72982d626136921dbR800) Create a utility for comparing if the app settings in SCM site match the ARM one, for ensuring the propagation is completed.
> 
>     3. [L820](https://github.com/Azure/azure-cli/pull/12262/files#diff-75a1767a60b87df72982d626136921dbR820) Create a REST request with retry pattern for querying app settings from the SCM site api/settings endpoint.
> 
> 
> If you have any further questions, feel free to let me know. Thanks in advance.

Thanks for detailed explanation.

**Comment by qwordy** [comment]
@Juliehzl hi, code owner, please take a look, thanks

**Comment by Hazhzeng** [comment]
Hi @Juliehzl, please take a look, this is scheduled to be to the S166 train.

**Pull request merged**